packager/packager.cc

```diff
@@ -0,0 +1,665 @@
+        return original;
+    }
+
+    ast::TreePtr preTransformReturn(core::MutableContext ctx, ast::TreePtr original) {
+        illegalControlFlow(ctx, original->loc, "`return`");
+        return original;
+    }
+
+    ast::TreePtr preTransformRescueCase(core::MutableContext ctx, ast::TreePtr original) {
+        illegalControlFlow(ctx, original->loc, "`rescue case`");
+        return original;
+    }
+
+    ast::TreePtr preTransformRescue(core::MutableContext ctx, ast::TreePtr original) {
+        illegalControlFlow(ctx, original->loc, "`rescue`");
+        return original;
+    }
+
+    ast::TreePtr preTransformAssign(core::MutableContext ctx, ast::TreePtr original) {
+        illegalControlFlow(ctx, original->loc, "`=`");
+        return original;
+    }
+
+    ast::TreePtr preTransformHash(core::MutableContext ctx, ast::TreePtr original) {
+        illegalControlFlow(ctx, original->loc, "hash literals");
+        return original;
+    }
+
+    ast::TreePtr preTransformArray(core::MutableContext ctx, ast::TreePtr original) {
+        illegalControlFlow(ctx, original->loc, "array literals");
+        return original;
+    }
+
+    ast::TreePtr preTransformMethodDef(core::MutableContext ctx, ast::TreePtr original) {
+        illegalControlFlow(ctx, original->loc, "method definitions");
+        return original;
+    }
+};
+
+// Sanity checks package files, mutates arguments to export / export_methods to point to item in namespace,
+// builds up the expression injected into packages that import the package, and codegens the <PackagedMethods>  module.
+shared_ptr<PackageInfo> getPackageInfo(core::MutableContext ctx, ast::ParsedFile &package) {
+    ENFORCE(package.file.exists());
+    ENFORCE(package.file.data(ctx).sourceType == core::File::Type::Package);
+    // Assumption: Root of AST is <root> class.
+    ENFORCE(static_cast<ast::ClassDef *>(package.tree.get()) != nullptr);
+    ENFORCE(static_cast<ast::ClassDef *>(package.tree.get())->symbol == core::Symbols::root());
+    auto packageFilePath = package.file.data(ctx).path();
+    ENFORCE(FileOps::getFileName(packageFilePath) == PACKAGE_FILE_NAME);
+    PackageInfoFinder finder;
+    package.tree = ast::TreeMap::apply(ctx, finder, move(package.tree));
+    finder.finalize(ctx);
+    if (finder.info) {
+        finder.info->packagePathPrefix = packageFilePath.substr(0, packageFilePath.find_last_of('/') + 1);
+    }
+    return move(finder.info);
+}
+
+// Add:
+//    module <PackageRegistry>::Mangled_Name_Package
+//      module ImportedPackage1
+//        # imported aliases go here
+//      end
+//    end
+// ...to __package.rb files to set up the package namespace.
+ast::ParsedFile rewritePackage(core::Context ctx, ast::ParsedFile file, const PackageInfo &package,
+                               const UnorderedMap<core::NameRef, shared_ptr<PackageInfo>> &packageInfoByMangledName) {
+    ast::ClassDef::RHS_store importedPackages;
+
+    // Sanity check: Packages _must_ be typed: strict
```

Comment:
I misread this at first as all code in packages, but I now realize you mean __package.rb spec files.

Reply:
Ah, I can edit this comment.